modin/core/execution/client/container.py

```diff
@@ -0,0 +1,742 @@
+        """
+        Read a CSV file from the specified filepath.
+
+        Parameters
+        ----------
+        connection : object
+            The data connection, e.g. a connnection to the database where the
+            service will store the result.
+        filepath : str
+            The filepath to read the CSV file from.
+        **kwargs : dict
+            Additional parameters to pass to the pandas read_csv function.
+
+        Returns
+        -------
+        UUID
+            The ID of the query compiler containing the read result.
+        """
+        io_result = self._io_class._read_csv(filepath, **kwargs)
+        if isinstance(io_result, self._query_compiler_class):
+            new_id = self._generate_id()
+            self._qc[new_id] = io_result
+            return new_id
+        return io_result
+
+    def read_sql(self, sql, connection, **kwargs) -> UUID:
+        """
+        Read data from a SQL connection.
+
+        Parameters
+        ----------
+        sql : str
+            SQL query to be executed or a table name.
+        connection : SQLAlchemy connectable, str, or sqlite3 connection
+            Using SQLAlchemy makes it possible to use any DB supported by that
+            library. If a DBAPI2 object, only sqlite3 is supported. The user is responsible
+            for engine disposal and connection closure for the SQLAlchemy
+            connectable; str connections are closed automatically. See
+            `here <https://docs.sqlalchemy.org/en/13/core/connections.html>`_.
+        **kwargs : dict
+            Parameters of ``read_sql`` function.
+
+        Returns
+        -------
+        UUID
+            ID of query compiler with data read in from SQL connection.
+        """
+        new_id = self._generate_id()
+        self._qc[new_id] = self._io_class._read_sql(sql, connection, **kwargs)
+        return new_id
+
+    def to_sql(self, id, **kwargs) -> None:
+        """
+        Write records stored in a DataFrame to a SQL database.
+
+        Databases supported by SQLAlchemy [1]_ are supported. Tables can be
+        newly created, appended to, or overwritten.
+
+        Parameters
+        ----------
+        id : UUID
+            ID of query compiler to write to database.
+        **kwargs : dict
+            Parameters of ``read_sql`` function.
+        """
+        self._io_class.to_sql(self._qc[id], **kwargs)
+
+
+def _set_forwarding_groupby_method(method_name: str):
+    """
+    Define a groupby method that forwards arguments to an inner query compiler.
+
+    Parameters
+    ----------
+    method_name : str
+    """
+
+    def forwarding_method(self, id, by_is_qc, by, *args, **kwargs):
+        if by_is_qc:
+            by = self._qc[by]
+        new_id = self._generate_id()
+        self._qc[new_id] = getattr(self._qc[id], method_name)(by, *args, **kwargs)
+        return new_id
+
+    setattr(ForwardingQueryCompilerContainer, method_name, forwarding_method)
```

Comment:
Just wanted to say that this is a super clever way of forwarding methods using `getattr` and `setattr`.

Reply:
+1. It saves a lot of lines of code!